### PR TITLE
[Merged by Bors] - depend on dioxus(and bevy)-maintained fork of stretch (taffy)

### LINKED ISSUE #677: very low performance when spawning UI nodes one inside the other
When spawning every second a new node as a children of the last one that had been spawned, the application starts to slow down after 10 iterations, and becomes unresponsive after 15~20 iterations

```rust
use bevy::prelude::*;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    App::build()
        .add_default_plugins()
        .add_startup_system(setup.system())
        .add_system(deep.system())
        .run();

    Ok(())
}

fn setup(mut commands: Commands) {
    commands.spawn(UiCameraComponents::default());

    commands
        .spawn(NodeComponents::default())
        .with(Timer::from_seconds(1., false));
}

fn deep(mut commands: Commands, entity: Entity, timer: &Timer) {
    if timer.just_finished {
        println!("spawning new child");
        commands
            .spawn(NodeComponents::default())
            .with(Timer::from_seconds(1., false));
        let child = commands.current_entity().unwrap();

        commands.push_children(entity, &[child]);
    }
}
```

with some logging and diagnostics:
```
Oct 14 00:19:36.576  INFO tank: spawning new child
Oct 14 00:19:36.577 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0145      (avg 0.0166)
Oct 14 00:19:36.578 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 60.3177     (avg 60.0176)
Oct 14 00:19:37.576 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0146      (avg 0.0165)
Oct 14 00:19:37.576 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 60.2317     (avg 60.1338)
Oct 14 00:19:37.592  INFO tank: spawning new child
Oct 14 00:19:38.580 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0178      (avg 0.0167)
Oct 14 00:19:38.580 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 60.0023     (avg 60.0193)
Oct 14 00:19:38.592  INFO tank: spawning new child
Oct 14 00:19:39.587 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0180      (avg 0.0163)
Oct 14 00:19:39.587 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 59.7169     (avg 60.2253)
Oct 14 00:19:39.602  INFO tank: spawning new child
Oct 14 00:19:40.589 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0174      (avg 0.0168)
Oct 14 00:19:40.590 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 59.9331     (avg 59.9481)
Oct 14 00:19:40.602  INFO tank: spawning new child
Oct 14 00:19:41.589 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0152      (avg 0.0166)
Oct 14 00:19:41.589 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 59.9631     (avg 59.9649)
Oct 14 00:19:41.621  INFO tank: spawning new child
Oct 14 00:19:42.573 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0159      (avg 0.0167)
Oct 14 00:19:42.573 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 60.3003     (avg 59.9682)
Oct 14 00:19:42.627  INFO tank: spawning new child
Oct 14 00:19:43.590 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0157      (avg 0.0166)
Oct 14 00:19:43.590 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 60.0817     (avg 60.0140)
Oct 14 00:19:43.644  INFO tank: spawning new child
Oct 14 00:19:44.584 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0172      (avg 0.0168)
Oct 14 00:19:44.585 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 59.9687     (avg 59.6242)
Oct 14 00:19:44.650  INFO tank: spawning new child
Oct 14 00:19:45.596 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0266      (avg 0.0263)
Oct 14 00:19:45.596 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 39.7056     (avg 38.1866)
Oct 14 00:19:45.664  INFO tank: spawning new child
Oct 14 00:19:46.628 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0447      (avg 0.0438)
Oct 14 00:19:46.628 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 34.7194     (avg 27.4654)
Oct 14 00:19:46.688  INFO tank: spawning new child
Oct 14 00:19:47.713 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0438      (avg 0.0656)
Oct 14 00:19:47.713 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 24.3330     (avg 20.1163)
Oct 14 00:19:47.722  INFO tank: spawning new child
Oct 14 00:19:48.738 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0511      (avg 0.0981)
Oct 14 00:19:48.738 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 22.6240     (avg 16.2551)
Oct 14 00:19:48.748  INFO tank: spawning new child
Oct 14 00:19:49.883 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0783      (avg 0.1336)
Oct 14 00:19:49.883 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 19.3643     (avg 13.7073)
Oct 14 00:19:49.893  INFO tank: spawning new child
Oct 14 00:19:51.104  INFO tank: spawning new child
Oct 14 00:19:52.276 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0769      (avg 0.1966)
Oct 14 00:19:52.276 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 17.5957     (avg 11.7911)
Oct 14 00:19:52.290  INFO tank: spawning new child
Oct 14 00:19:54.523 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0798      (avg 0.2521)
Oct 14 00:19:54.523 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 17.0662     (avg 11.1097)
Oct 14 00:19:54.540  INFO tank: spawning new child
Oct 14 00:19:58.888 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0784      (avg 0.3606)
Oct 14 00:19:58.888 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 16.5496     (avg 10.3950)
Oct 14 00:19:58.911  INFO tank: spawning new child
Oct 14 00:20:07.772 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0784      (avg 0.5752)
Oct 14 00:20:07.772 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 16.0953     (avg 9.6545)
Oct 14 00:20:07.807  INFO tank: spawning new child
Oct 14 00:20:26.534 DEBUG bevy_log_diagnostic: diagnostic: frame_time               : 0.0793      (avg 1.0161)
Oct 14 00:20:26.534 DEBUG bevy_log_diagnostic: diagnostic: fps                      : 15.6552     (avg 8.8989)
Oct 14 00:20:26.615  INFO tank: spawning new child
```
the FPS starts to slow down after 10 seconds, and is false after 15 seconds as some seconds are skipped.

### PR DESCRIPTION
**Pull request by colepoirier**
# Objective

DioxusLabs and Bevy have taken over maintaining what was our abandoned ui layout dependency [stretch](https://github.com/vislyhq/stretch). Dioxus' fork has had a lot of work done on it by @alice-i-cecile, @Weibye , @jkelleyrtp, @mockersf, @HackerFoo, @TimJentzsch and a dozen other contributors and now is in much better shape than stretch was. The updated crate is called taffy and is available on github [here](https://github.com/DioxusLabs/taffy) ([taffy](https://crates.io/crates/taffy) on crates.io). The goal of this PR is to replace stretch v0.3.2 with taffy v0.1.0.

## Solution

I changed the bevy_ui Cargo.toml to depend on taffy instead of stretch and fixed all the errors rustc complained about.

---

## Changelog

Changed bevy_ui layout dependency from stretch to taffy (the maintained fork of stretch).

fixes #677

## Migration Guide

The public api of taffy is different from that of stretch so please advise me on what to do here @alice-i-cecile.

### CODE REVIEW

**Comment by colepoirier** [comment]
## Bevy UI Examples Screenshots

Here are the screenshots for each of the five bevy ui examples showing that the layout is broken by the changes in this PR. 

### text_debug

#### main

main_example_text_debug

#### this PR

stretch2_example_text_debug

### text

#### main

main_example_text

#### this PR

stretch2_example_text

### font_atlas_debug

#### main

main_example_font_atlas_debug

#### this PR

stretch2_example_font_atlas_debug

### button

#### main

main_example_button

#### this PR

stretch2_example_button

### ui

#### main

main_example_ui

#### this PR

stretch2_example_ui

**Comment by alice-i-cecile** [comment]
Those breaks are pretty rough :(

**Comment by rparrett** [comment]
It would be really helpful if Dioxus maintained a changelog. It seems like the "24 commits ahead" are:

- Renaming the project
- https://github.com/DioxusLabs/stretch/pull/1
- https://github.com/DioxusLabs/stretch/pull/2

So just a couple of layout bug fixes so far.

**Comment by colepoirier** [comment]
> It would be really helpful if Dioxus maintained a changelog. It seems like the "24 commits ahead" are:
> 
>     * Renaming the project
> 
>     * [Jk/nested fix DioxusLabs/stretch#1](https://github.com/DioxusLabs/stretch/pull/1)
> 
>     * [fix re-computing layout moves children of non-zero positioned parent DioxusLabs/stretch#2](https://github.com/DioxusLabs/stretch/pull/2)
> 
> 
> So just a couple of layout bug fixes so far.
> 
> Are bevy's examples relying on buggy layout behavior, or are these fixes introducing new bugs? Or has the migration to stretch2 gone awry in some way?

All excellent points @rparrett!

**Comment by colepoirier** [comment]
[Jk/nested fix DioxusLabs/stretch#1](https://github.com/DioxusLabs/stretch/pull/1) includes 17 commits FYI

**Comment by Sheepyhead** [comment]
I tested this PR with my most complicated piece of bevy_ui code; an inventory and equipment half-window. Here is what it looks like on Bevy main at time of writing:
image
And the following is what it looks like when I depend on this PR's changes:
image
The code has been extracted from a game I've been working on and is largely unchanged except for the sake of brevity and cutting out all game-specific mechanics, you can see the repo here https://github.com/Sheepyhead/bevy_ui_test

**Comment by mockersf** [comment]
Stretch used to give relative layout for nodes, now it's the absolute layout. I believe it's an issue on Stretch side and opened https://github.com/DioxusLabs/stretch/pull/6, but it's not complicated to fix on Bevy side if they want to stick to the new behaviour.

**Comment by colepoirier** [comment]
> Stretch used to give relative layout for nodes, now it's the absolute layout. I believe it's an issue on Stretch side and opened [DioxusLabs/stretch#6](https://github.com/DioxusLabs/stretch/pull/6), but it's not complicated to fix on Bevy side if they want to stick to the new behaviour.

Tested the bevy ui and button examples using François' PR fixing relative positioning and it looks like it fixes it.

#### ui

mockersf_stretch2_example_ui

#### button

mockersf_stretch2_example_button

**Comment by Sheepyhead** [comment]
I can confirm that using Francois' latest changes on this PR completely fixes the layout of my complicated gui 
image

**Comment by alice-i-cecile** [comment]
Excellent; I'll see about cutting a release of stretch2 ASAP then.

**Comment by colepoirier** [comment]
The two UI examples look correct using stretch2 v0.4.3.

### ui

stretch2_0 4 3_example_ui

### button

stretch2_0 4 3_example_button

@mockersf Are there any further steps I need to take with this PR?

**Comment by mockersf** [comment]
> @mockersf Are there any further steps I need to take with this PR?

I think @alice-i-cecile intend to release the crate renamed soon, maybe wait for it

**Comment by mockersf** [comment]
> Edit: It looks like the check-bans CI is still failing on something to do with the lockfile that I don't understand. Do you know what is happening here and if I need to do something to resolve it?

It seems the new crate has a few dependencies in multiple versions which we don't allow in Bevy. You can investigate on the dependencies if the duplicates are really needed, then open PRs on the dependencies to update if possible... It can take a lot of time

**Comment by alice-i-cecile** [comment]
> You can investigate on the dependencies if the duplicates are really needed, then open PRs on the dependencies to update if possible... It can take a lot of time

I will happily accept any PRs to do this for `sprawl`. I'll bump the version of the dependencies where I can for the initial release of `sprawl` as well.

**Comment by Weibye** [comment]
> It seems the new crate has a few dependencies in multiple versions which we don't allow in Bevy. You can investigate on the dependencies if the duplicates are really needed, then open PRs on the dependencies to update if possible... It can take a lot of time

Seems all dependency issues start with `heapless` (direct dependency in sprawl) and it's dependency tree

**Comment by colepoirier** [comment]
> > It seems the new crate has a few dependencies in multiple versions which we don't allow in Bevy. You can investigate on the dependencies if the duplicates are really needed, then open PRs on the dependencies to update if possible... It can take a lot of time
> 
> Seems all dependency issues start with `heapless` (direct dependency in sprawl) and it's dependency tree

I'm not sure how to fix this. Do you have any suggestions?

**Comment by alice-i-cecile** [comment]
@colepoirier can you make a PR to `sprawl` removing `heapless`? I don't think it's pulling its weight, and I would like to simplify that section of the code anyways.

**Comment by colepoirier** [comment]
> @colepoirier can you make a PR to `sprawl` removing `heapless`? I don't think it's pulling its weight, and I would like to simplify that section of the code anyways.

@alice-i-cecile switched from stretch2 v0.4.3 to sprawl git main to check if that resolves the CI check-bans failures. Regardless I will attempt to remove heapless from sprawl and make a PR. Additionally, the sprawl v 0.1.0 release seems to be missing the vast majority of the contents of the library so will need a v 0.2.0 before we can take a dependency on it for bevy.

**Comment by colepoirier** [comment]
Removing heapless dependency from sprawl allows this to finally pass the check-bans CI :)

**Comment by colepoirier** [comment]
sprawl was renamed to taffy and v0.1.0 was realeased on crates.io today!

File: crates/bevy_ui/src/flex/mod.rs [hunk 1]
```diff
@@ -14,32 +14,32 @@ use bevy_transform::components::Transform;
...
-use stretch::{number::Number, Stretch};
+use taffy::{number::Number, Taffy};
 
 pub struct FlexSurface {
-    entity_to_stretch: HashMap<Entity, stretch::node::Node>,
-    window_nodes: HashMap<WindowId, stretch::node::Node>,
-    stretch: Stretch,
+    entity_to_sprawl: HashMap<Entity, taffy::node::Node>,
+    window_nodes: HashMap<WindowId, taffy::node::Node>,
+    sprawl: Taffy,
```
Comment (Weibye) [inline crates/bevy_ui/src/flex/mod.rs:22]: This should be `taffy`, not `sprawl`
```rust
    entity_to_taffy: HashMap<Entity, taffy::node::Node>,
    window_nodes: HashMap<WindowId, taffy::node::Node>,
    taffy: Taffy,
```

File: crates/bevy_ui/src/flex/mod.rs [hunk 2]
```diff
@@ -14,32 +14,32 @@ use bevy_transform::components::Transform;
...
 }
 
-// SAFE: as long as MeasureFunc is Send + Sync. https://github.com/vislyhq/stretch/issues/69
+// SAFE: as long as MeasureFunc is Send + Sync. https://github.com/vislyhq/sprawl/issues/69
```
Comment (Weibye) [inline crates/bevy_ui/src/flex/mod.rs:25]:
```suggestion
// SAFE: as long as MeasureFunc is Send + Sync. https://github.com/vislyhq/stretch/issues/69
```
This should refer to the old issue in the original `stretch` repo

File: crates/bevy_ui/src/flex/mod.rs [hunk 3]
```diff
@@ -14,32 +14,32 @@ use bevy_transform::components::Transform;
...
 
 fn _assert_send_sync_flex_surface_impl_safe() {
     fn _assert_send_sync<T: Send + Sync>() {}
-    _assert_send_sync::<HashMap<Entity, stretch::node::Node>>();
-    _assert_send_sync::<HashMap<WindowId, stretch::node::Node>>();
-    // FIXME https://github.com/vislyhq/stretch/issues/69
-    // _assert_send_sync::<Stretch>();
+    _assert_send_sync::<HashMap<Entity, taffy::node::Node>>();
+    _assert_send_sync::<HashMap<WindowId, taffy::node::Node>>();
+    // FIXME https://github.com/vislyhq/sprawl/issues/69
```
Comment (Weibye) [inline crates/bevy_ui/src/flex/mod.rs:35]:
```suggestion
    // FIXME https://github.com/vislyhq/stretch/issues/69
```
same her with the link

File: crates/bevy_ui/src/flex/mod.rs [hunk 4]
```diff
@@ -14,32 +14,32 @@ use bevy_transform::components::Transform;
...
+    // _assert_send_sync::<sprawl>();
```
Comment (Weibye) [inline crates/bevy_ui/src/flex/mod.rs:36]:
```suggestion
    // _assert_send_sync::<Taffy>();
```

File: crates/bevy_ui/src/flex/mod.rs [hunk 5]
```diff
@@ -14,32 +14,32 @@ use bevy_transform::components::Transform;
...
 }
 
 impl fmt::Debug for FlexSurface {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         f.debug_struct("FlexSurface")
-            .field("entity_to_stretch", &self.entity_to_stretch)
+            .field("entity_to_sprawl", &self.entity_to_sprawl)
```
Comment (Weibye) [inline crates/bevy_ui/src/flex/mod.rs:42]: See comment above, rename these

File: crates/bevy_ui/src/flex/mod.rs [hunk 6]
```diff
@@ -48,27 +48,25 @@ impl fmt::Debug for FlexSurface {
...
     }
 }
 
 impl FlexSurface {
     pub fn upsert_node(&mut self, entity: Entity, style: &Style, scale_factor: f64) {
         let mut added = false;
-        let stretch = &mut self.stretch;
-        let stretch_style = convert::from_style(scale_factor, style);
-        let stretch_node = self.entity_to_stretch.entry(entity).or_insert_with(|| {
+        let sprawl = &mut self.sprawl;
```
Comment (Weibye) [inline crates/bevy_ui/src/flex/mod.rs:61]: I'll not put suggestions on all of these, but these should be `sprawl` -> `taffy` before merging

**Review by Weibye — CHANGES_REQUESTED** [review]

File: crates/bevy_ui/src/flex/mod.rs (hunk 2 above)
Comment (Weibye) [reply crates/bevy_ui/src/flex/mod.rs:25]: Or maybe we rather should create a copy of the issue in Taffy and link to it there? @alice-i-cecile

File: crates/bevy_ui/src/flex/mod.rs (hunk 6 above)
Comment (colepoirier) [reply crates/bevy_ui/src/flex/mod.rs]: @Weibye thanks for the heads up! I've fixed them all I think :)

**Comment by alice-i-cecile** [comment]
@colepoirier the PR title and PR description need to be updated too.

**Comment by colepoirier** [comment]
> @colepoirier the PR title and PR description need to be updated too.

I rewrote it thanks for the heads up! @alice-i-cecile what should I do about the removal of `Direction` from taffy in bevy_ui? What I did is just delete the code that used to be `impl From<Direction> for stretch::style::Direction`. Is that the right thing to do? Is `Direction` something that will need to be added back to bevy_ui after it is re-implemented in taffy?

File: crates/bevy_ui/src/flex/mod.rs [hunk 7]
```diff
@@ -33,7 +33,7 @@ fn _assert_send_sync_flex_surface_impl_safe() {
     _assert_send_sync::<HashMap<Entity, taffy::node::Node>>();
     _assert_send_sync::<HashMap<WindowId, taffy::node::Node>>();
     // FIXME https://github.com/vislyhq/taffy/issues/69
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/flex/mod.rs]: a) is this FIXME comment still relevant and b) the link is wrong

**Review by Weibye — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
`Direction` in this sense refers to text direction. IMO it's out of scope for a layout library. Bevy should implement its own type when it needs this.

File: crates/bevy_ui/src/flex/mod.rs (hunk 2 above)
Comment (alice-i-cecile) [reply crates/bevy_ui/src/flex/mod.rs]: Oh right. IMO we should just remove MeasureFunc here: https://github.com/DioxusLabs/taffy/issues/101

File: crates/bevy_ui/src/flex/mod.rs [hunk 8]
```diff
@@ -30,16 +30,16 @@ unsafe impl Sync for FlexSurface {}
 
 fn _assert_send_sync_flex_surface_impl_safe() {
     fn _assert_send_sync<T: Send + Sync>() {}
-    _assert_send_sync::<HashMap<Entity, stretch::node::Node>>();
-    _assert_send_sync::<HashMap<WindowId, stretch::node::Node>>();
+    _assert_send_sync::<HashMap<Entity, taffy::node::Node>>();
+    _assert_send_sync::<HashMap<WindowId, taffy::node::Node>>();
     // FIXME https://github.com/vislyhq/stretch/issues/69
```
Comment (colepoirier) [inline crates/bevy_ui/src/flex/mod.rs:35]: @alice-i-cecile the FIXME link is fixed on the latest commit. I'm unsure why github hasn't marked that as outdated?

File: crates/bevy_ui/src/flex/mod.rs (hunk 2 above)
Comment (colepoirier) [reply crates/bevy_ui/src/flex/mod.rs]: What should I do for this PR?

**Review by alice-i-cecile — APPROVED** [review]
That was about as painless as I could have hoped. @colepoirier have you verified that the UI examples all look okay with the newest changes?

File: crates/bevy_ui/src/flex/mod.rs (hunk 2 above)
Comment (alice-i-cecile) [reply crates/bevy_ui/src/flex/mod.rs]: Create a new issue on Taffy and link to that.

Comment (colepoirier) [reply crates/bevy_ui/src/flex/mod.rs]: Done https://github.com/DioxusLabs/taffy/issues/146

**Comment by colepoirier** [comment]
> That was about as painless as I could have hoped. @colepoirier have you verified that the UI examples all look okay with the newest changes?

Yes both look identical to the screenshots above for the bevy_ui on main.

#### ui

taffy_example_button

#### button

taffy_example_ui

**Comment by colepoirier** [comment]
FYI I also removed the section for stretch in the deny.toml. I think that lies within the scope of this PR but I'd like confirmation. Please let me know if that is correct.

**Comment by TimJentzsch** [comment]
> `Direction` in this sense refers to text direction. IMO it's out of scope for a layout library. Bevy should implement its own type when it needs this.

Since the flexbox algorithm depends on writing direction to determine start and end we will probably need to add a way to provide this again.

**Comment by mockersf** [comment]
just tried a complex case that was running at 45 fps with stretch, I get 160 fps with this PR 👍

**Review by cart — APPROVED** [review]
Looks good to me! The Bevy side was an easy review. Couldn't find any regressions. Did a taffy/stretch diff review and the changes look good to me.

I'm excited for our new future of collaborative UI layout dev!

**Pull request closed without merging**